File: src/services/code_action/autofix_match_syntax.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type kind =
  | ObjShorthandToConst
  | ObjShorthandToReference
  | InvalidMatchStatementBody
  | InvalidBindingKind
  | InvalidWildcardSyntax
  | InvalidCaseSyntax
  | NonExhaustiveObjectPattern of {
      add_rest: bool;
      missing_props: string list;
    }
  | NotExhaustive of (Loc.t, Loc.t) Flow_ast.MatchPattern.t list
  | UnusedPattern

let is_valid_ident_name = Parser_flow.string_is_valid_identifier_name

let is_invalid_case_syntax invalid_syntax =
  let {
    Flow_ast.Match.Case.InvalidSyntax.invalid_prefix_case;
    invalid_infix_colon;
    invalid_suffix_semicolon;
  } =
    invalid_syntax
  in
  Base.Option.is_some invalid_prefix_case
  || Base.Option.is_some invalid_infix_colon
  || Base.Option.is_some invalid_suffix_semicolon

let empty_invalid_case_syntax =
  {
    Flow_ast.Match.Case.InvalidSyntax.invalid_prefix_case = None;
    invalid_infix_colon = None;
    invalid_suffix_semicolon = None;
  }

class mapper target_loc ~kind =
  object (this)
    inherit Flow_ast_contains_mapper.mapper target_loc as super

    method! match_object_pattern_property prop =
      let open Flow_ast.MatchPattern.ObjectPattern in
      let (loc, _) = prop in
      if not @@ this#is_target loc then
        super#match_object_pattern_property prop
      else
        match prop with
        | (_, Property.Valid _) -> super#match_object_pattern_property prop
        | (_, Property.InvalidShorthand id) ->
          let key = Property.Identifier id in
          (match kind with
          | ObjShorthandToConst ->
            let pattern =
              ( Loc.none,
                Flow_ast.MatchPattern.BindingPattern
                  {
                    Flow_ast.MatchPattern.BindingPattern.kind = Flow_ast.Variable.Const;
                    id;
                    comments = None;
                  }
              )
            in
            (Loc.none, Property.Valid { Property.key; pattern; shorthand = true; comments = None })
          | ObjShorthandToReference ->
            let pattern = (Loc.none, Flow_ast.MatchPattern.IdentifierPattern id) in
            (Loc.none, Property.Valid { Property.key; pattern; shorthand = false; comments = None })
          | _ -> super#match_object_pattern_property prop)

    method! match_statement loc x =
      let { Flow_ast.Match.match_keyword_loc; _ } = x in
      match kind with
      | InvalidMatchStatementBody when this#target_contained_by loc ->
        let on_case_body stmt =
          let (stmt_loc, _) = stmt in
          let body =
            if not @@ this#is_target stmt_loc then
              stmt
            else
              ( Loc.none,
                Flow_ast.Statement.Block { Flow_ast.Statement.Block.body = [stmt]; comments = None }
              )
          in
          this#statement body
        in
        this#match_ loc ~on_case_body x
      | NotExhaustive patterns_to_add when this#is_target match_keyword_loc ->
        let { Flow_ast.Match.arg; cases; match_keyword_loc; comments } = x in
        let new_cases =
          Base.List.map patterns_to_add ~f:(fun pattern ->
              let body =
                ( Loc.none,
                  Flow_ast.Statement.Block { Flow_ast.Statement.Block.body = []; comments = None }
                )
              in
              ( Loc.none,
                {
                  Flow_ast.Match.Case.pattern;
                  body;
                  guard = None;
                  comments = None;
                  invalid_syntax = empty_invalid_case_syntax;
                }
              )
          )
        in
        let x = { Flow_ast.Match.arg; cases = cases @ new_cases; match_keyword_loc; comments } in
        super#match_statement loc x
      | _ -> super#match_statement loc x

    method! match_expression loc x =
      let { Flow_ast.Match.match_keyword_loc; _ } = x in
      match kind with
      | NotExhaustive patterns_to_add when this#is_target match_keyword_loc ->
        let { Flow_ast.Match.arg; cases; match_keyword_loc; comments } = x in
        let new_cases =
          Base.List.map patterns_to_add ~f:(fun pattern ->
              let body =
                ( Loc.none,
                  Flow_ast.Expression.Identifier
                    (Loc.none, { Flow_ast.Identifier.name = "undefined"; comments = None })
                )
              in
              ( Loc.none,
                {
                  Flow_ast.Match.Case.pattern;
                  body;
                  guard = None;
                  comments = None;
                  invalid_syntax = empty_invalid_case_syntax;
                }
              )
          )
        in
        let x = { Flow_ast.Match.arg; cases = cases @ new_cases; match_keyword_loc; comments } in
        super#match_expression loc x
      | _ -> super#match_expression loc x

    method! match_pattern pattern =
      let open Flow_ast.MatchPattern in
      let pattern =
        match (pattern, kind) with
        | ((loc, BindingPattern binding), InvalidBindingKind) when this#is_target loc ->
          let { BindingPattern.kind = _; id; comments } = binding in
          (loc, BindingPattern { BindingPattern.kind = Flow_ast.Variable.Const; id; comments })
        | ((loc, WildcardPattern wildcard), InvalidWildcardSyntax) when this#is_target loc ->
          let { WildcardPattern.comments; invalid_syntax_default_keyword = _ } = wildcard in
          (loc, WildcardPattern { WildcardPattern.comments; invalid_syntax_default_keyword = false })
        | ( (loc, ObjectPattern { ObjectPattern.properties; rest; comments }),
            NonExhaustiveObjectPattern { add_rest; missing_props }
          )
          when this#is_target loc ->
          let rest =
            if add_rest then
              Some (Loc.none, { RestPattern.argument = None; comments = None })
            else
              rest
          in
          let properties =
            if Base.List.is_empty missing_props then
              properties
            else
              let added_props =
                Base.List.map missing_props ~f:(fun name ->
                    let key =
                      if is_valid_ident_name name then
                        ObjectPattern.Property.Identifier
                          (Loc.none, { Flow_ast.Identifier.name; comments = None })
                      else
                        ObjectPattern.Property.StringLiteral
                          ( Loc.none,
                            {
                              Flow_ast.StringLiteral.value = name;
                              raw = "" (* Raw unused *);
                              comments = None;
                            }
                          )
                    in
                    let pattern =
                      ( Loc.none,
                        WildcardPattern
                          {
                            WildcardPattern.comments = None;
                            invalid_syntax_default_keyword = false;
                          }
                      )
                    in
                    ( Loc.none,
                      ObjectPattern.Property.Valid
                        { ObjectPattern.Property.key; pattern; shorthand = false; comments = None }
                    )
                )
              in
              properties @ added_props
          in
          (loc, ObjectPattern { ObjectPattern.properties; rest; comments })
        | ( (loc, ObjectPattern { ObjectPattern.properties; rest = Some (rest_loc, _); comments }),
            UnusedPattern
          )
          when this#is_target rest_loc ->
          (loc, ObjectPattern { ObjectPattern.properties; rest = None; comments })
        | ( (loc, ArrayPattern { ArrayPattern.elements; rest = Some (rest_loc, _); comments }),
            UnusedPattern
          )
          when this#is_target rest_loc ->
          (loc, ArrayPattern { ArrayPattern.elements; rest = None; comments })
        | ((loc, OrPattern { OrPattern.patterns; comments }), UnusedPattern)
          when this#target_contained_by loc ->
          let (patterns_rev, changed) =
            Base.List.fold patterns ~init:([], false) ~f:(fun (patterns_rev, changed) pattern ->
                let (pattern_loc, _) = pattern in
                if this#is_target pattern_loc then
                  (patterns_rev, true)
                else if changed then
                  (* Update loc for subsequent patterns so that no newline appears. *)
                  let pattern = (Loc.none, snd pattern) in
                  (pattern :: patterns_rev, changed)
                else
                  (pattern :: patterns_rev, changed)
            )
          in
          if changed then
            match patterns_rev with
            | [single_pattern] -> single_pattern
            | _ -> (loc, OrPattern { OrPattern.patterns = Base.List.rev patterns_rev; comments })
          else
            pattern
        | _ -> pattern
      in
      super#match_pattern pattern

    method! match_ loc ~on_case_body x =
      let open Flow_ast.Match in
      let { match_keyword_loc; _ } = x in
      let x =
        match kind with
        | InvalidCaseSyntax when this#is_target match_keyword_loc ->
          let { arg; cases; match_keyword_loc; comments } = x in
          let cases =
            Base.List.map cases ~f:(fun case ->
                let (loc, { Case.pattern; body; guard; comments; invalid_syntax }) = case in
                if is_invalid_case_syntax invalid_syntax then
                  let invalid_syntax = empty_invalid_case_syntax in
                  (loc, { Case.pattern; body; guard; comments; invalid_syntax })
                else
                  case
            )
          in
          { arg; cases; match_keyword_loc; comments }
        | UnusedPattern when this#target_contained_by loc ->
          let { arg; cases; match_keyword_loc; comments } = x in
          let (cases_rev, changed) =
            Base.List.fold cases ~init:([], false) ~f:(fun (cases_rev, changed) case ->
                let (_, { Case.pattern = (pattern_loc, _); _ }) = case in
                if this#is_target pattern_loc then
                  (cases_rev, true)
                else if changed then
                  (* Update loc for subsequent cases so that no newline appears. *)
                  let case = (Loc.none, snd case) in
                  (case :: cases_rev, changed)
                else
                  (case :: cases_rev, changed)
            )
          in
          if changed then
            { arg; cases = Base.List.rev cases_rev; match_keyword_loc; comments }
          else
            x
        | _ -> x
      in
      super#match_ loc ~on_case_body x

    method! match_case ~on_case_body case =
      let open Flow_ast.Match.Case in
      let (loc, { pattern; body; guard; comments; invalid_syntax }) = case in
      let case =
        if
          kind = InvalidCaseSyntax
          && this#target_contained_by loc
          && is_invalid_case_syntax invalid_syntax
        then
          let invalid_syntax = empty_invalid_case_syntax in
          (loc, { pattern; body; guard; comments; invalid_syntax })
        else
          case
      in
      super#match_case ~on_case_body case
  end

let convert_object_shorthand_to_const ast loc =
  let mapper = new mapper loc ~kind:ObjShorthandToConst in
  mapper#program ast

let convert_object_shorthand_to_reference ast loc =
  let mapper = new mapper loc ~kind:ObjShorthandToReference in
  mapper#program ast

let fix_invalid_match_statement_body ast loc =
  let mapper = new mapper loc ~kind:InvalidMatchStatementBody in
  mapper#program ast

let fix_invalid_binding_kind ast loc =
  let mapper = new mapper loc ~kind:InvalidBindingKind in
  mapper#program ast

let fix_invalid_wildcard_syntax ast loc =
  let mapper = new mapper loc ~kind:InvalidWildcardSyntax in
  mapper#program ast

let fix_invalid_case_syntax ast loc =
  let mapper = new mapper loc ~kind:InvalidCaseSyntax in
  mapper#program ast

let fix_non_exhaustive_object_pattern ~add_rest missing_props ast loc =
  let mapper = new mapper loc ~kind:(NonExhaustiveObjectPattern { add_rest; missing_props }) in
  mapper#program ast

let fix_not_exhaustive patterns_to_add ast loc =
  let mapper = new mapper loc ~kind:(NotExhaustive patterns_to_add) in
  mapper#program ast

let remove_unused_pattern ast loc =
  let mapper = new mapper loc ~kind:UnusedPattern in
  mapper#program ast

File: rust_port/crates/flow_services_code_action/src/autofix_match_syntax.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::sync::Arc;

use dupe::Dupe;
use flow_parser::ast;
use flow_parser::ast::IdentifierInner;
use flow_parser::ast::VariableKind;
use flow_parser::ast::expression::ExpressionInner;
use flow_parser::ast::match_;
use flow_parser::ast::match_pattern;
use flow_parser::ast::match_pattern::object_pattern;
use flow_parser::ast::statement;
use flow_parser::ast::statement::StatementInner;
use flow_parser::ast_visitor::AstVisitor;
use flow_parser::ast_visitor::map_match_default;
use flow_parser::ast_visitor::map_match_expression_default;
use flow_parser::ast_visitor::map_match_object_pattern_property_default;
use flow_parser::ast_visitor::map_match_pattern_default;
use flow_parser::ast_visitor::map_match_statement_default;
use flow_parser::ast_visitor::map_program_default;
use flow_parser::ast_visitor::map_statement_default;
use flow_parser::js_id_unicode;
use flow_parser::loc::LOC_NONE;
use flow_parser::loc::Loc;

use crate::contains_mapper::ContainsMapper;

pub enum Kind {
    ObjShorthandToConst,
    ObjShorthandToReference,
    InvalidMatchStatementBody,
    InvalidBindingKind,
    InvalidWildcardSyntax,
    InvalidCaseSyntax,
    NonExhaustiveObjectPattern {
        add_rest: bool,
        missing_props: Vec<String>,
    },
    NotExhaustive(Vec<match_pattern::MatchPattern<Loc, Loc>>),
    UnusedPattern,
}

fn is_valid_ident_name(name: &str) -> bool {
    js_id_unicode::string_is_valid_identifier_name(name)
}

fn is_invalid_case_syntax(invalid_syntax: &match_::InvalidSyntax<Loc>) -> bool {
    let match_::InvalidSyntax {
        invalid_prefix_case,
        invalid_infix_colon,
        invalid_suffix_semicolon,
    } = invalid_syntax;
    invalid_prefix_case.is_some()
        || invalid_infix_colon.is_some()
        || invalid_suffix_semicolon.is_some()
}

fn empty_invalid_case_syntax() -> match_::InvalidSyntax<Loc> {
    match_::InvalidSyntax {
        invalid_prefix_case: None,
        invalid_infix_colon: None,
        invalid_suffix_semicolon: None,
    }
}

struct Mapper {
    contains: ContainsMapper,
    kind: Kind,
    fix_all_invalid_case_syntax: bool,
}

impl<'ast> AstVisitor<'ast, Loc> for Mapper {
    fn normalize_loc(loc: &Loc) -> &Loc {
        loc
    }

    fn normalize_type(type_: &Loc) -> &Loc {
        type_
    }

    fn map_match_object_pattern_property(
        &mut self,
        prop: &object_pattern::Property<Loc, Loc>,
    ) -> object_pattern::Property<Loc, Loc> {
        let loc = prop.loc();
        if !self.contains.is_target(loc) {
            return map_match_object_pattern_property_default(self, prop);
        }
        match prop {
            object_pattern::Property::Valid { .. } => {
                map_match_object_pattern_property_default(self, prop)
            }
            object_pattern::Property::InvalidShorthand {
                loc: _,
                identifier: id,
            } => {
                let key = object_pattern::Key::Identifier(id.dupe());
                match &self.kind {
                    Kind::ObjShorthandToConst => {
                        let pattern = match_pattern::MatchPattern::BindingPattern {
                            loc: LOC_NONE,
                            inner: Arc::new(match_pattern::BindingPattern {
                                kind: VariableKind::Const,
                                id: id.dupe(),
                                comments: None,
                            }),
                        };
                        object_pattern::Property::Valid {
                            loc: LOC_NONE,
                            property: object_pattern::PropertyStruct {
                                key,
                                pattern,
                                shorthand: true,
                                comments: None,
                            },
                        }
                    }
                    Kind::ObjShorthandToReference => {
                        let pattern = match_pattern::MatchPattern::IdentifierPattern {
                            loc: LOC_NONE,
                            inner: id.dupe(),
                        };
                        object_pattern::Property::Valid {
                            loc: LOC_NONE,
                            property: object_pattern::PropertyStruct {
                                key,
                                pattern,
                                shorthand: false,
                                comments: None,
                            },
                        }
                    }
                    _ => map_match_object_pattern_property_default(self, prop),
                }
            }
        }
    }

    fn map_match_statement(
        &mut self,
        loc: &Loc,
        m: &statement::MatchStatement<Loc, Loc>,
    ) -> statement::MatchStatement<Loc, Loc> {
        let match_keyword_loc = &m.match_keyword_loc;
        match &self.kind {
            Kind::InvalidMatchStatementBody if self.contains.target_contained_by(loc) => self
                .map_match_(loc, m, |visitor, stmt| {
                    let stmt_loc = stmt.loc();
                    if !visitor.contains.is_target(stmt_loc) {
                        visitor.map_statement(stmt)
                    } else {
                        let wrapped = statement::Statement::new(StatementInner::Block {
                            loc: LOC_NONE,
                            inner: Arc::new(statement::Block {
                                body: Arc::from([stmt.dupe()]),
                                comments: None,
                            }),
                        });
                        visitor.map_statement(&wrapped)
                    }
                }),
            Kind::NotExhaustive(patterns_to_add) if self.contains.is_target(match_keyword_loc) => {
                let match_::Match {
                    arg,
                    cases,
                    match_keyword_loc,
                    comments,
                } = m;
                let new_cases: Vec<match_::Case<Loc, Loc, statement::Statement<Loc, Loc>>> =
                    patterns_to_add
                        .iter()
                        .map(|pattern| {
                            let body = statement::Statement::new(StatementInner::Block {
                                loc: LOC_NONE,
                                inner: Arc::new(statement::Block {
                                    body: Arc::from([]),
                                    comments: None,
                                }),
                            });
                            match_::Case {
                                loc: LOC_NONE,
                                pattern: pattern.clone(),
                                body,
                                guard: None,
                                comments: None,
                                invalid_syntax: empty_invalid_case_syntax(),
                                case_match_root_loc: LOC_NONE,
                            }
                        })
                        .collect();
                let mut all_cases: Vec<_> = cases.iter().cloned().collect();
                all_cases.extend(new_cases);
                let new_m = match_::Match {
                    arg: arg.dupe(),
                    cases: Arc::from(all_cases),
                    match_keyword_loc: match_keyword_loc.dupe(),
                    comments: comments.clone(),
                };
                map_match_statement_default(self, loc, &new_m)
            }
            _ => map_match_statement_default(self, loc, m),
        }
    }

    fn map_match_expression(
        &mut self,
        loc: &Loc,
        m: &ast::expression::MatchExpression<Loc, Loc>,
    ) -> ast::expression::MatchExpression<Loc, Loc> {
        let match_keyword_loc = &m.match_keyword_loc;
        match &self.kind {
            Kind::NotExhaustive(patterns_to_add) if self.contains.is_target(match_keyword_loc) => {
                let match_::Match {
                    arg,
                    cases,
                    match_keyword_loc,
                    comments,
                } = m;
                let new_cases: Vec<match_::Case<Loc, Loc, ast::expression::Expression<Loc, Loc>>> =
                    patterns_to_add
                        .iter()
                        .map(|pattern| {
                            let body =
                                ast::expression::Expression::new(ExpressionInner::Identifier {
                                    loc: LOC_NONE,
                                    inner: ast::Identifier::new(IdentifierInner {
                                        loc: LOC_NONE,
                                        name: "undefined".into(),
                                        comments: None,
                                    }),
                                });
                            match_::Case {
                                loc: LOC_NONE,
                                pattern: pattern.clone(),
                                body,
                                guard: None,
                                comments: None,
                                invalid_syntax: empty_invalid_case_syntax(),
                                case_match_root_loc: LOC_NONE,
                            }
                        })
                        .collect();
                let mut all_cases: Vec<_> = cases.iter().cloned().collect();
                all_cases.extend(new_cases);
                let new_m = match_::Match {
                    arg: arg.dupe(),
                    cases: Arc::from(all_cases),
                    match_keyword_loc: match_keyword_loc.dupe(),
                    comments: comments.clone(),
                };
                map_match_expression_default(self, loc, &new_m)
            }
            _ => map_match_expression_default(self, loc, m),
        }
    }

    fn map_match_pattern(
        &mut self,
        pattern: &match_pattern::MatchPattern<Loc, Loc>,
    ) -> match_pattern::MatchPattern<Loc, Loc> {
        let pattern = match (pattern, &self.kind) {
            (
                match_pattern::MatchPattern::BindingPattern {
                    loc,
                    inner: binding,
                },
                Kind::InvalidBindingKind,
            ) if self.contains.is_target(loc) => {
                let match_pattern::BindingPattern {
                    kind: _,
                    id,
                    comments,
                } = &**binding;
                match_pattern::MatchPattern::BindingPattern {
                    loc: loc.dupe(),
                    inner: Arc::new(match_pattern::BindingPattern {
                        kind: VariableKind::Const,
                        id: id.dupe(),
                        comments: comments.clone(),
                    }),
                }
            }
            (
                match_pattern::MatchPattern::WildcardPattern {
                    loc,
                    inner: wildcard,
                },
                Kind::InvalidWildcardSyntax,
            ) if self.contains.is_target(loc) => {
                let match_pattern::WildcardPattern {
                    comments,
                    invalid_syntax_default_keyword: _,
                } = wildcard;
                match_pattern::MatchPattern::WildcardPattern {
                    loc: loc.dupe(),
                    inner: match_pattern::WildcardPattern {
                        comments: comments.clone(),
                        invalid_syntax_default_keyword: false,
                    },
                }
            }
            (
                match_pattern::MatchPattern::ObjectPattern { loc, inner: obj },
                Kind::NonExhaustiveObjectPattern {
                    add_rest,
                    missing_props,
                },
            ) if self.contains.is_target(loc) => {
                let match_pattern::ObjectPattern {
                    properties,
                    rest,
                    comments,
                } = &**obj;
                let rest = if *add_rest {
                    Some(match_pattern::RestPattern {
                        loc: LOC_NONE,
                        argument: None,
                        comments: None,
                    })
                } else {
                    rest.clone()
                };
                let properties = if missing_props.is_empty() {
                    properties.to_vec()
                } else {
                    let added_props: Vec<object_pattern::Property<Loc, Loc>> = missing_props
                        .iter()
                        .map(|name| {
                            let key = if is_valid_ident_name(name) {
                                object_pattern::Key::Identifier(ast::Identifier::new(
                                    IdentifierInner {
                                        loc: LOC_NONE,
                                        name: name.as_str().into(),
                                        comments: None,
                                    },
                                ))
                            } else {
                                object_pattern::Key::StringLiteral((
                                    LOC_NONE,
                                    ast::StringLiteral {
                                        value: name.as_str().into(),
                                        raw: "".into(),
                                        comments: None,
                                    },
                                ))
                            };
                            let pattern = match_pattern::MatchPattern::WildcardPattern {
                                loc: LOC_NONE,
                                inner: match_pattern::WildcardPattern {
                                    comments: None,
                                    invalid_syntax_default_keyword: false,
                                },
                            };
                            object_pattern::Property::Valid {
                                loc: LOC_NONE,
                                property: object_pattern::PropertyStruct {
                                    key,
                                    pattern,
                                    shorthand: false,
                                    comments: None,
                                },
                            }
                        })
                        .collect();
                    let mut result = properties.to_vec();
                    result.extend(added_props);
                    result
                };
                match_pattern::MatchPattern::ObjectPattern {
                    loc: loc.dupe(),
                    inner: Arc::new(match_pattern::ObjectPattern {
                        properties: Arc::from(properties),
                        rest,
                        comments: comments.clone(),
                    }),
                }
            }
            (
                match_pattern::MatchPattern::ObjectPattern { loc, inner: obj },
                Kind::UnusedPattern,
            ) if obj
                .rest
                .as_ref()
                .is_some_and(|r| self.contains.is_target(&r.loc)) =>
            {
                let match_pattern::ObjectPattern {
                    properties,
                    rest: _,
                    comments,
                } = &**obj;
                match_pattern::MatchPattern::ObjectPattern {
                    loc: loc.dupe(),
                    inner: Arc::new(match_pattern::ObjectPattern {
                        properties: properties.dupe(),
                        rest: None,
                        comments: comments.clone(),
                    }),
                }
            }
            (
                match_pattern::MatchPattern::ArrayPattern { loc, inner: arr },
                Kind::UnusedPattern,
            ) if arr
                .rest
                .as_ref()
                .is_some_and(|r| self.contains.is_target(&r.loc)) =>
            {
                let match_pattern::ArrayPattern {
                    elements,
                    rest: _,
                    comments,
                } = &**arr;
                match_pattern::MatchPattern::ArrayPattern {
                    loc: loc.dupe(),
                    inner: Arc::new(match_pattern::ArrayPattern {
                        elements: elements.dupe(),
                        rest: None,
                        comments: comments.clone(),
                    }),
                }
            }
            (
                match_pattern::MatchPattern::OrPattern { loc, inner: or_pat },
                Kind::UnusedPattern,
            ) if self.contains.target_contained_by(loc) => {
                let match_pattern::OrPattern { patterns, comments } = &**or_pat;
                let (mut patterns_rev, changed) = patterns.iter().fold(
                    (Vec::new(), false),
                    |(mut patterns_rev, changed), pat| {
                        let pattern_loc = pat.loc();
                        if self.contains.is_target(pattern_loc) {
                            (patterns_rev, true)
                        } else if changed {
                            let mut new_pat = pat.clone();
                            *new_pat.loc_mut() = LOC_NONE;
                            patterns_rev.push(new_pat);
                            (patterns_rev, changed)
                        } else {
                            patterns_rev.push(pat.clone());
                            (patterns_rev, changed)
                        }
                    },
                );
                if changed {
                    patterns_rev.reverse();
                    match patterns_rev.len() {
                        1 => patterns_rev.into_iter().next().unwrap(),
                        _ => match_pattern::MatchPattern::OrPattern {
                            loc: loc.dupe(),
                            inner: Arc::new(match_pattern::OrPattern {
                                patterns: Arc::from(patterns_rev),
                                comments: comments.clone(),
                            }),
                        },
                    }
                } else {
                    pattern.clone()
                }
            }
            _ => pattern.clone(),
        };
        map_match_pattern_default(self, &pattern)
    }

    fn map_match_<B, B2>(
        &mut self,
        loc: &'ast Loc,
        m: &'ast match_::Match<Loc, Loc, B>,
        mut on_case_body: impl FnMut(&mut Self, &'ast B) -> B2,
    ) -> match_::Match<Loc, Loc, B2>
    where
        Loc: Dupe,
    {
        let match_keyword_loc = &m.match_keyword_loc;
        match &self.kind {
            Kind::InvalidCaseSyntax if self.contains.is_target(match_keyword_loc) => {
                self.fix_all_invalid_case_syntax = true;
                let result = map_match_default(self, loc, m, on_case_body);
                self.fix_all_invalid_case_syntax = false;
                result
            }
            Kind::UnusedPattern if self.contains.target_contained_by(loc) => {
                let match_::Match {
                    arg,
                    cases,
                    match_keyword_loc,
                    comments,
                } = m;
                let mut keep_info: Vec<(usize, bool)> = Vec::new(); // (index, reset_loc)
                let mut changed = false;
                for (i, case) in cases.iter().enumerate() {
                    let pattern_loc = case.pattern.loc();
                    if self.contains.is_target(pattern_loc) {
                        changed = true;
                    } else if changed {
                        keep_info.push((i, true));
                    } else {
                        keep_info.push((i, false));
                    }
                }
                if changed {
                    let arg_ = self.map_expression(arg);
                    let cases_: Arc<[match_::Case<Loc, Loc, B2>]> = keep_info
                        .iter()
                        .map(|&(idx, reset_loc)| {
                            let mut case_b2 = self.map_match_case(&cases[idx], &mut on_case_body);
                            if reset_loc {
                                case_b2.loc = LOC_NONE;
                            }
                            case_b2
                        })
                        .collect();
                    let comments_ = self.map_syntax_opt(comments.as_ref());
                    match_::Match {
                        arg: arg_,
                        cases: cases_,
                        match_keyword_loc: match_keyword_loc.dupe(),
                        comments: comments_,
                    }
                } else {
                    map_match_default(self, loc, m, on_case_body)
                }
            }
            _ => map_match_default(self, loc, m, on_case_body),
        }
    }

    fn map_match_case<B, B2>(
        &mut self,
        case: &'ast match_::Case<Loc, Loc, B>,
        on_case_body: &mut impl FnMut(&mut Self, &'ast B) -> B2,
    ) -> match_::Case<Loc, Loc, B2>
    where
        Loc: Dupe,
    {
        let match_::Case {
            loc,
            pattern,
            body,
            guard,
            comments,
            invalid_syntax,
            case_match_root_loc,
        } = case;
        let use_empty_invalid_syntax = matches!(self.kind, Kind::InvalidCaseSyntax)
            && (self.fix_all_invalid_case_syntax || self.contains.target_contained_by(loc))
            && is_invalid_case_syntax(invalid_syntax);
        let pattern_ = self.map_match_pattern(pattern);
        let body_ = on_case_body(self, body);
        let guard_ = guard.as_ref().map(|g| self.map_expression(g));
        let comments_ = self.map_syntax_opt(comments.as_ref());
        match_::Case {
            loc: loc.dupe(),
            pattern: pattern_,
            body: body_,
            guard: guard_,
            comments: comments_,
            invalid_syntax: if use_empty_invalid_syntax {
                empty_invalid_case_syntax()
            } else {
                invalid_syntax.clone()
            },
            case_match_root_loc: case_match_root_loc.dupe(),
        }
    }

    fn map_program(&mut self, program: &ast::Program<Loc, Loc>) -> ast::Program<Loc, Loc> {
        if self.contains.should_map_program(program) {
            map_program_default(self, program)
        } else {
            program.clone()
        }
    }

    fn map_statement(
        &mut self,
        stmt: &ast::statement::Statement<Loc, Loc>,
    ) -> ast::statement::Statement<Loc, Loc> {
        if self.contains.should_map_statement(stmt) {
            map_statement_default(self, stmt)
        } else {
            stmt.dupe()
        }
    }
}

pub fn convert_object_shorthand_to_const(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::ObjShorthandToConst,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn convert_object_shorthand_to_reference(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::ObjShorthandToReference,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_invalid_match_statement_body(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InvalidMatchStatementBody,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_invalid_binding_kind(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InvalidBindingKind,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_invalid_wildcard_syntax(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InvalidWildcardSyntax,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_invalid_case_syntax(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InvalidCaseSyntax,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_non_exhaustive_object_pattern(
    add_rest: bool,
    missing_props: Vec<String>,
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::NonExhaustiveObjectPattern {
            add_rest,
            missing_props,
        },
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn fix_not_exhaustive(
    patterns_to_add: Vec<match_pattern::MatchPattern<Loc, Loc>>,
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::NotExhaustive(patterns_to_add),
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}

pub fn remove_unused_pattern(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::UnusedPattern,
        fix_all_invalid_case_syntax: false,
    };
    mapper.map_program(ast)
}
